### PR TITLE
[c10d] Implement `__instancecheck__` for `c10d::ReduceOp`

### PR DESCRIPTION
**Pull request by crcrpar**
Summary:
- Customize the metaclass of `torch.distributed.distributed_c10d.ReduceOp` for the sake of custom `__instancecheck__`
- Add `copy.copy`, `copy.deepcopy`, and `pickle` support with tests

Rel:
- #81272
- #84243
- #87191
- #87303
- #87555

Ref:
- https://github.com/pybind/pybind11/issues/2696

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @wanchaol

### CODE REVIEW

**Review by wanchaol — APPROVED** [review]
Looks good! Thanks for proactively figuring out a solution for a better ReduceOp! Have two minor comments.

File: torch/csrc/distributed/c10d/init.cpp [hunk 1]
```diff
@@ -562,9 +604,51 @@ This class does not support ``__members__`` property.)");
...
+              return py::make_tuple(r.op_, py::none());
+            }
+            TORCH_CHECK(r.supplement_.defined(), "Invalid PREMUL_SUM ReduceOp");
+            const auto* preMulSupplement =
+                reinterpret_cast<::c10d::NCCLPreMulSumSupplement*>(
+                    r.supplement_.get());
+            if (preMulSupplement->tensor_factors.empty()) {
+              return py::make_tuple(r.op_, preMulSupplement->double_factor);
+            } else {
+              return py::make_tuple(r.op_, preMulSupplement->tensor_factors);
```
[the remark this answers is not in the document]
Comment (crcrpar) [reply torch/csrc/distributed/c10d/init.cpp:646]: let me think about...

[the remark this answers is not in the document]
Comment (crcrpar) [reply torch/csrc/distributed/c10d/init.cpp]: @wanchaol  would you mind taking another look?
30ab4bc96982f8c68b1fc00620ee2b2fb9492a4a changes to a Tensor

**Review by wanchaol — APPROVED** [review]
lgtm, one nit.

**Pull request closed without merging**